### PR TITLE
fix: Resolve Algolia DocSearch duplicate results and search bar layout

### LINKED ISSUE #22607: [Docs] Feedback: SignalR Docs Tutorial/Sample
### On which page? 📄

https://platform.uno/docs/articles/signalr.html

### What's wrong? ❗

To me its unclear, how this Sample/Tutorial exactly shows to me the integration in a Uno App, because it seems to only show ASP NET Core Programm.cs code, but nothing I would know from my regular Uno App Template.

The Tutorial should be improved by adding a Uno App context, like Supported Platforms and what we can expect in the end to see. The current registration in asp net core makes me as new to SignalR-User assume, that:
- I will end up with Razor Pages in my Uno App Solution instead of getting a Uno Wasm target driven UI? (see the `MapRazorPages()` call in the second code sample block contained in this docs page!)
- The Questions:
  - "Does Uno or at least its browserwasm target behave exactly like Razor Pages?"
  - "So am I able to register and show my Uno Wasm UI on a ASP NET Core Server to be shown at runtime?"
  - "Is the `Uno.Wasm.Bootstrap.Server` NuGet of Uno possibly required for exactly this?" since I only have been able to find docs for non-Server package using samples in the Bootstrapper docs (*not* equivalent to the Uno App Templates contained Server option, I am aware of this! But this is exactly the missing point!)

The UnoChat Sample Readme in Uno.Samples Repo does compared to this docs page show to me a (Web?) UI and Console app, while I would not know of any way to have a Console App as UI in Uno (I indeed tryed this, but was not possible to establish using the current docs. nobody tells what hostbuilder setup in programm.cs I would be needing. And also the sample global.json file shows to me a 5.3.96 version, which is defintly not a Version I would rely on, when searching for a lookup. too many versions and especially breaking changes (SkiaRenderer, Desktop target...) did happen meanwhile.

additionally, it would make sense, to let the reader know of all of the code we seem to need in the ViewModel, if I would want to beleave the Sample eitherway:
https://github.com/unoplatform/Uno.Samples/blob/master/UI/ChatSignalR/src/UnoChat/Presentation/ViewModel.cs
and turns out, the answer to the filed question above with the "will I end up using Razor Pages in a Uno App as UI instead of a Uno Wasm target" is "Yes, you will rather use Razor then Web Assembly UI build with the Uno.Sdk!"

Image

### Any feedback? 💬

Just a feedback from my POV, this tutorial paints to me a really sad image and impression when it comes to the Uno Wasm Target integration.

I know that we can not have a server side rendering, but is our Wasm target really this kind of bad integrate-able with a ASP NET Core Server, so we need to make advertisement for sending Uno Users over to using Razor / Blazor instead, where *it seems to be possible to integrate and fully/detailed document this* with a Server project? 🤷 

I have been searching for a Gprc using sample, like the Maui Client of the dotnet/eShop Sample is doing, but while I am quite sure there have been one in the Uno docs in the past, all I can find now is the Migration page here: https://platform.uno/docs/articles/guides/silverlight-migration/08-data-access-overview.html#_content of silverlight, ***which projects has been removed from VS support up from the 2017 Version!!!***

https://learn.microsoft.com/en-us/visualstudio/releases/2026/port-migrate-and-upgrade-visual-studio-projects
> <html>
> <body>
> 
> Silverlight | Silverlight projects were removed in Visual Studio 2017.
> -- | --
> 
> </body>
> </html>

## Child issues
- [ ] https://github.com/unoplatform/Uno.Samples/issues/910

### LINKED ISSUE #22788: fix: Algolia DocSearch results are duplicated and search bar spans full viewport width
## Current behavior

The Algolia DocSearch integration in the docs site has several issues:

1. **Duplicate search results**: The DocSearch crawler indexes breadcrumb navigation, heading-level anchors, and same-titled pages from different sections, producing duplicate or near-duplicate results.
2. **Search bar spans full viewport width**: When the search modal opens, the search input bar stretches across the entire screen instead of staying within the modal. This is caused by docfx's default CSS rule `header { position: fixed; top: 0; left: 0; }` unintentionally applying to DocSearch's `<header class="DocSearch-SearchBar">` element.
3. **Search not visible in local builds**: The custom TOC template (`toc.html.tmpl`) that renders the DocSearch button is silently ignored by newer docfx versions, which expect `toc.html.primary.tmpl`.

## Expected behavior

- Search results should be unique and deduplicated
- The search bar should stay within the modal bounds (~560px)
- Search should be visible and functional in local doc builds

## How to reproduce

1. Visit the docs site and open DocSearch (Cmd+K)
2. Search for "MVUX" or any common term
3. Observe duplicate results and the search input spanning the full viewport width

## Workaround

None

## Works on

All platforms (browser-based docs site)

## Issues found and fixed

1. **Breadcrumb duplicates** — The crawler indexed `<div id="breadcrumb">` content as separate search records with `#breadcrumb` anchors. Fixed by filtering those in `transformItems` and adding `data-docsearch-exclude` to the breadcrumb template.

2. **Same-page heading duplicates** — DocSearch creates one record per heading level, so a single page could appear multiple times. Fixed with URL-based deduplication that keeps only the first hit per base page URL.

3. **Same-titled different pages** — Pages like workshop variants shared identical titles, making results look like duplicates. Fixed by detecting same-titled groups and appending a distinguishing URL path segment to the display text (updating both `_snippetResult` and `_highlightResult` so DocSearch actually renders it).

4. **Navigation noise in index** — The crawler was indexing navbar, footer, and sidebar content. Fixed by adding `data-docsearch-exclude` to those elements via new template overrides (`affix.tmpl.partial`, `footer.tmpl.partial`, `breadcrumb.tmpl.partial`, and `_navbar.tmpl.partial`).

5. **Search bar spanning full viewport width** — docfx's default CSS `header { position: fixed; top: 0; left: 0; }` was applied to DocSearch's `<header class="DocSearch-SearchBar">`, pulling it out of the 560px modal. Fixed with a CSS override in `main.css`.

6. **Search not visible in local builds** — The custom TOC template was named `toc.html.tmpl` but docfx 2.73+ expects `toc.html.primary.tmpl`, so the `#docsearch` container was silently missing. Fixed by renaming the file and pinning the docfx version.

### PR DESCRIPTION
**Pull request by Jen-Uno**
## What this PR does

Fixes the Algolia DocSearch problems reported in #22788 — duplicate / low-quality results, a search bar that spans the full viewport width, and search not showing up in local builds — and re-scopes the fixes to the correct layer per the analysis in `unoplatform/uno-private#2038`.

## Approach (layered)

DocSearch renders whatever records live in the **hosted** Algolia index, so duplicates, navigation noise and ranking are fundamentally an **index/crawler-side** concern — not something client JS can fully fix. This PR splits the work accordingly:

- **Index / crawler side (the real fix):** exclusion, per-page dedup, URL canonicalization and ranking, expressed as a crawler configuration to apply to the hosted `platform` index.
- **Display side (interim polish + UX):** a small, unit-tested `transformItems` module, the search-bar layout fix, and a "See all results" entry point to a full results page.

> Note: `data-docsearch-exclude` HTML attributes are **not** honored by the DocSearch crawler (verified against Algolia's docs) — exclusion must be server-side. The template markers remain as intent documentation, but the authoritative exclusion is the crawler config below.

### Index/crawler configuration — the authoritative fix (new)
- `doc/algolia/docsearch-crawler-config.js` — `recordExtractor` with `.remove()` for navbar/footer/sidebar/breadcrumb, `attributeForDistinct: 'url_without_anchor'` + `distinct` for one-record-per-page, URL canonicalization (`/index.html`, trailing slash), a `section` facet for same-title disambiguation, and `customRanking`.
- `doc/algolia/README.md` — how to apply it to the hosted index, **staging-index validation**, and acceptance checks. (The crawler write key is a placeholder — never committed.)

### Client-side result transforms — refactored & unit-tested
- `doc/templates/uno/service/docsearch-transform.js` — extracts the former inline `transformItems` logic (breadcrumb filtering, page-level dedup, same-title disambiguation) into a pure, testable module. Fixes a **prototype-pollution crash** (a page titled `constructor`/`__proto__` previously threw and killed `transformItems`) and the **short-first-URL** disambiguation edge case.
- `doc/templates/uno/service/docsearch-transform.test.js` — `node --test` coverage (13 tests) incl. the prototype-key and short-URL cases.
- `scripts.tmpl.partial` delegates to the module with an identity fallback so search never hard-breaks.

### "See all results" + full results page
- Replaced the previous custom ~250-line `MutationObserver` overlay (and its `algoliasearch-lite` dependency) with DocSearch's built-in **`resultsFooterComponent`**.
- `doc/search.html` — a self-contained **InstantSearch.js** results page (query in URL, pagination, highlighting) the footer links to. Uses only the public search-only key. Wired into `docfx.json` resources.

### Search-bar layout fix (kept)
- `service/search.scss` — override for docfx's global `header { position: fixed }` leaking onto `.DocSearch-SearchBar`, keeping it inside the modal (+ small footer-link styling).

### Build / local compatibility
- `gulpfile.js` — exclude `*.test.js` from the JS bundle (test files are Node-only and would otherwise break `docfx.js`).
- `import_external_docs_test.ps1` — pin docfx `2.73.2` to match CI; `sidebar.js`/`docfx.js` keep the `toc.json → toc.html` compatibility guard.

## Testing
- `cd doc/templates/uno/service && node --test` → 13/13 passing.
- The changed files were adversarially reviewed (crawler API, `resultsFooterComponent` v3 signature, InstantSearch usage, docfx/gulp wiring); findings fixed.

## Smoke-test before merge (needs the live widget)
- The v3 `resultsFooterComponent` returns a Preact vnode from a separately-loaded Preact; it fails safe (no footer) if the runtime rejects it — confirm it renders on the real modal.
- The footer count depends on `state.context.nbHits`; the label falls back to "See all results" if absent.
- The crawler config must be applied to the hosted index (on a **staging index first**) — that's where the duplicate/ranking fixes actually take effect.

## Not in this PR
- The SignalR tutorial modernization that was previously bundled here has been split into **#23992** (`Fixes #22607`), as it's unrelated to search.

Fixes #22788 · Refs `unoplatform/uno-private#2038`

### CODE REVIEW

**Comment by Jen-Uno** [comment]
Audit note: I accidentally resolved three review threads while working in the wrong PR context, then immediately reopened them. Their status is now unchanged and still unresolved for normal review handling.

**Review by ajpinedam — APPROVED** [review]

File: doc/import_external_docs_test.ps1 [hunk 1]
```diff
@@ -11,7 +11,31 @@ param(
 )
 
 Write-Host 'Updating docfx tool...' -ForegroundColor Black -BackgroundColor Green
-dotnet tool update --global docfx
+# Intentionally pinned to match the CI docs build toolchain (build/Uno.UI.Build.csproj)
+# so local validation reproduces doc generation behavior consistently.
+dotnet tool update --global docfx --version 2.73.2 --allow-downgrade
```
[the remark this answers is not in the document]
Comment (Jen-Uno) [reply doc/import_external_docs_test.ps1:16]: By design — search visibility in local builds is restored via the `toc.json → toc.html` sidebar compatibility guard (`sidebar.js` / `docfx.js`) together with the docfx `2.73.2` pin, so `toc.html.tmpl` is intentionally **not** renamed to `toc.html.primary.tmpl`. Rationale in `unoplatform/uno-private#2038` (the runtime guard supersedes the template rename).

[the remark this answers is not in the document]
Comment (Jen-Uno) [reply doc/import_external_docs_test.ps1:16]: Addressed — the script now falls back to `dotnet tool install --global docfx --version 2.73.2` when `dotnet tool update` returns a non-zero exit code (i.e. docfx isn't installed on a clean machine).

File: doc/import_external_docs_test.ps1 [hunk 2]
```diff
@@ -11,7 +11,31 @@ param(
...
+# Older or newer local versions can cause template behavior differences.
+# Verify the actual resolved version and force-install the pinned toolchain when needed.
+# A Homebrew or other system-wide install may shadow the dotnet tool, so prefer
+# the dotnet tool path explicitly when it exists.
+$dotnetToolDocfx = Join-Path $env:HOME '.dotnet/tools/docfx'
+if (Test-Path $dotnetToolDocfx) {
+    $script:DocfxCmd = $dotnetToolDocfx
+} else {
+    $script:DocfxCmd = 'docfx'
+}
```
[the remark this answers is not in the document]
Comment (Jen-Uno) [reply doc/import_external_docs_test.ps1]: Addressed — the dotnet global tool is now resolved via `[Environment]::GetFolderPath('UserProfile')` instead of `$env:HOME` (which is unset on Windows PowerShell) and checks for `docfx.exe` as well as `docfx`, re-resolving the path after any reinstall.

File: doc/articles/signalr.md [hunk 3]
```diff
@@ -2,51 +2,275 @@
...
+[ASP.NET Core SignalR](https://learn.microsoft.com/aspnet/core/signalr/introduction) is an open-source library that simplifies adding real-time web functionality to apps. It enables server-side code to push content to connected clients instantly using WebSockets, Server-Sent Events, or Long Polling (chosen automatically based on capabilities).
+
+SignalR is a good fit for Uno Platform apps that need:
+
+- Real-time messaging or chat
+- Live dashboards and monitoring
+- Collaborative editing
+- Push notifications from a server
+
+This guide walks you through creating an ASP.NET Core SignalR server and connecting to it from an Uno Platform client app. The SignalR .NET client (`Microsoft.AspNetCore.SignalR.Client`) works on all Uno target platforms — WebAssembly, Desktop (Windows, macOS, Linux via Skia), iOS, and Android.
```
[the remark this answers is not in the document]
Comment (Jen-Uno) [reply doc/articles/signalr.md]: Split out — the SignalR tutorial modernization now lives in #23992 (`Fixes #22607`) and has been removed from this branch, keeping this PR scoped to DocSearch.

File: doc/search.html [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+      function hitTitle(hit) {
+        var h = hit._highlightResult && hit._highlightResult.hierarchy;
+        var raw = hit.hierarchy || {};
+        function lvl(n) {
+          var key = 'lvl' + n;
+          return (h && h[key] && h[key].value) || raw[key];
+        }
+        return lvl(6) || lvl(5) || lvl(4) || lvl(3) || lvl(2) || lvl(1) || lvl(0) || 'Untitled';
+      }
```
[the remark this answers is not in the document]
Comment (Jen-Uno) [reply doc/search.html:115]: Addressed in 5aa7be9 — the raw `hit.hierarchy` / `hit.content` fallbacks in `hitTitle()`/`hitSnippet()` are now passed through `esc()`. The connector-escaped highlight/snippet values (which carry the intended `<mark>` markup) are left as-is.

File: doc/search.html [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+        // that one so the full parent trail is kept even for lvl4–lvl6 titles.
+        parts.pop();
+        return parts.join(' › ');
+      }
+
+      function hitSnippet(hit) {
+        var s = hit._snippetResult && hit._snippetResult.content && hit._snippetResult.content.value;
+        if (!s) { s = hit.content || ''; if (s.length > 200) s = s.slice(0, 200) + '…'; }
+        return s;
+      }
```
[the remark this answers is not in the document]
Comment (Jen-Uno) [reply doc/search.html:133]: Addressed in 5aa7be9 — the raw `hit.hierarchy` / `hit.content` fallbacks in `hitTitle()`/`hitSnippet()` are now passed through `esc()`. The connector-escaped highlight/snippet values (which carry the intended `<mark>` markup) are left as-is.

File: doc/search.html [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+          templates: {
+            empty: '<div class="uno-empty">No results{{#query}} for “{{query}}”{{/query}}.</div>',
+            item: function (hit) {
+              var crumb = hitCrumb(hit);
+              return '<a class="uno-hit" href="' + esc(hit.url) + '">'
+                + (crumb ? '<div class="uno-hit-crumb">' + esc(crumb) + '</div>' : '')
+                + '<div class="uno-hit-title">' + hitTitle(hit) + '</div>'
+                + '<div class="uno-hit-snippet">' + hitSnippet(hit) + '</div>'
+                + '</a>';
+            }
```
[the remark this answers is not in the document]
Comment (Jen-Uno) [reply doc/search.html:168]: Addressed in 5aa7be9 — the raw `hit.hierarchy` / `hit.content` fallbacks in `hitTitle()`/`hitSnippet()` are now passed through `esc()`. The connector-escaped highlight/snippet values (which carry the intended `<mark>` markup) are left as-is.

File: doc/search.html [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+          }
+        }),
+
+        w.pagination({ container: '#pagination' })
+      ]);
```
[the remark this answers is not in the document]
Comment (Jen-Uno) [reply doc/search.html:173]: Addressed — added a `safeUrl()` that allowlists `http:`/`https:`/relative/anchor URLs (strips control chars, maps `javascript:`/`data:`/etc. to `#`); `hit.url` now goes through `safeUrl()` before `esc()` into the href.
